opentracing-shim/src/opentracingshim/__init__.py

```diff
@@ -0,0 +1,241 @@
+        # TODO: Return a saved instance of SpanWrapper instead of constructing
+        # a new object (and the same for ScopeWrapper?).
+        # https://github.com/open-telemetry/opentelemetry-python/issues/161#issuecomment-534136274
+
+
+class TracerWrapper(opentracing.Tracer):
+    def __init__(self, tracer, scope_manager=None):
+        self._otel_tracer = tracer
+
+        # If a scope manager isn't provided by the user, create a
+        # `ScopeManagerWrapper` instance and use it to initialize the
+        # `TracerWrapper`.
+        if scope_manager is None:
+            scope_manager = ScopeManagerWrapper(self)
+
+        super().__init__(scope_manager=scope_manager)
+
+    def unwrap(self):
+        """Returns the wrapped OpenTelemetry `Tracer` object."""
+
+        return self._otel_tracer
+
+    @property
+    def active_span(self):
+        span = self._otel_tracer.get_current_span()
+        if span is None:
+            return None
+        return SpanWrapper(self, span.get_context(), span)
+        # TODO: Return a saved instance of SpanWrapper instead of constructing
+        # a new object.
+        # https://github.com/open-telemetry/opentelemetry-python/issues/161#issuecomment-534136274
+
+    @contextmanager
+    def start_active_span(
+        self,
+        operation_name,
+        child_of=None,
+        references=None,
+        tags=None,
+        start_time=None,
+        ignore_active_span=False,
+        finish_on_close=True,
+    ):
+        span = self.start_span(
+            operation_name=operation_name,
+            child_of=child_of,
+            references=references,
+            tags=tags,
+            start_time=start_time,
+            ignore_active_span=ignore_active_span,
+        )
+        with self._scope_manager.activate(span, finish_on_close) as scope:
+            yield scope
+
+    def start_span(
+        self,
+        operation_name=None,
+        child_of=None,
+        references=None,
+        tags=None,
+        start_time=None,
+        ignore_active_span=False,
+    ):
+        parent = child_of
+        if parent is not None:
+            if isinstance(parent, SpanWrapper):
+                parent = child_of.unwrap()
+            elif isinstance(parent, SpanContextWrapper):
+                parent = child_of.unwrap()
+            else:
+                raise RuntimeError(
+                    "Invalid parent type when calling start_span()."
+                )
+            # TODO: Should we use the OpenTracing base classes for the type
+            # check?
+
+        # Use active span as parent when no explicit parent is specified.
+        if (
+            self.active_span is not None
+            and not ignore_active_span
+            and not parent
+        ):
+            parent = self.active_span.unwrap()
+
+        span = self._otel_tracer.create_span(operation_name, parent)
+
+        if references:
+            if not isinstance(references, list):
```

Comment:
That's dangerous. What if I pass references in a tuple? Is it really allowed to pass a single reference without wrapping it in some Iterable?

Reply:
This is no excuse, but I stole it from https://github.com/jaegertracing/jaeger-client-python/blob/8155a31ae9ecbb42f91212342bcc7fc2c1a33016/jaeger_client/tracer.py#L161-L162 :-)

The OpenTracing API specified a list: https://github.com/opentracing/opentracing-python/blob/7d2e62b5999d3cdec9a9c2720a34a1d11f425461/opentracing/tracer.py#L103

Shall we simply drop this check?

Reply:
Both good points! `opentracing-python` really does specify a list here, and it seems reasonable to expect that this shim will behave the same way as the jaeger client.

I don't see any downside to replacing the instance check with `iter` (or a try/catch around `for ref...`), but up to you if you want to remove the check altogether.

One argument for keeping the check is that we should be [minimizing unhandled exceptions](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/error-handling.md), but this isn't something we've paid close attention to so far. On the other hand, it would seem to satisfy the OpenTracing API to raise here if `references` is the wrong type.

Reply:
Yeah, the as-a-list handling should be fine (because of the OT requirements). We can add handling for other objects (such as tuples) to be on the very safe side, but that would be a plus.

Reply:
Thanks everybody for the feedback. I'll remove the check for now and allow things to break in case the wrong type is passed.
True, we can be extra forgiving here, but then the question is - why just here? In theory we could try to deal with wrong types everywhere, which kind of defeats the purpose of having a well-defined contract IMO. The argument name is plural (`references`) and the API docs are very clear about the type. Let's leave it at that.

Reply:
You didn't remove the check though?

Reply:
I've removed it in https://github.com/open-telemetry/opentelemetry-python/pull/211/commits/79bd31a642cbcd08a8d21bb02d50a058b65090d9.